mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/mapping/building/impl/PojoMapper.java

```diff
@@ -172,35 +179,41 @@ private void prepareEntityOrIndexedType(PojoRawTypeModel<?> rawTypeModel,
 	}
 
 	@Override
-	public void mapIndexedTypes(IndexManagerBuildingStateProvider indexManagerBuildingStateProvider) {
+	public void mapIndexedTypes(IndexedEntityBindingContextProvider contextProvider) {
 		for ( PojoRawTypeModel<?> indexedEntityType : indexedEntityTypes ) {
 			try {
-				mapIndexedType( indexedEntityType, indexManagerBuildingStateProvider );
+				mapIndexedType( indexedEntityType, contextProvider );
 			}
 			catch (RuntimeException e) {
 				failureCollector.withContext( EventContexts.fromType( indexedEntityType ) )
 						.add( e );
 			}
 		}
+
+		if ( !failureCollector.hasFailure() ) {
+			checkPathTrackers();
+		}
 	}
 
 	private void mapIndexedType(PojoRawTypeModel<?> indexedEntityType,
-			IndexManagerBuildingStateProvider indexManagerBuildingStateProvider) {
+			IndexedEntityBindingContextProvider contextProvider) {
 		PojoTypeAdditionalMetadata metadata = typeAdditionalMetadataProvider.get( indexedEntityType );
 		// This metadata is guaranteed to exist; see prepareEntityOrIndexedType()
 		PojoIndexedTypeAdditionalMetadata indexedTypeMetadata = metadata.getIndexedTypeMetadata().get();
 		PojoEntityTypeAdditionalMetadata entityTypeMetadata = metadata.getEntityTypeMetadata().get();
 
 		IndexManagerBuildingState<?> indexManagerBuildingState =
-				indexManagerBuildingStateProvider.getIndexManagerBuildingState(
+				contextProvider.getIndexManagerBuildingState(
 						indexedTypeMetadata.getBackendName(),
 						indexedTypeMetadata.getIndexName()
 								.orElse( entityTypeMetadata.getEntityName() ),
 						multiTenancyEnabled
 				);
-
+		IndexedEntityBindingContext bindingContext = contextProvider.createIndexedEntityBindingContext(
+				this, indexManagerBuildingState.getSchemaRootNodeBuilder()
+		);
```

Comment:
IMHO, collaborations here have become too complex. The readability is not the very good.
I think that is very important having code that is very understandable.

If the goal here is to create an instance of `PojoIndexedTypeManagerBuilder`, maybe it is better to do all the job in `indexManagerBuildingState`. 
In general, I like to have responsible objects, instead of having object providers which unique responsibility is to provide other instances.
As a rule of thumb, try to write sequence diagram. These never lie. If these are simple, collaborations are good.

Do you think we could simplify a bit the collaborations here? Thanks

Reply:
> If the goal here is to create an instance of PojoIndexedTypeManagerBuilder, maybe it is better to do all the job in indexManagerBuildingState.

Well as the names indicate, `PojoIndexedTypeManagerBuilder` is implemented in the POJO module. `indexManagerBuildingState`, on the other hand, is implemented in the engine and mainly delegates to the backends.
There are multiple components at work here, all of which can have multiple implementations, and each requiring a different execution context. So I cannot just stuff everything into a single component to "simplify" collaboration.

I agree this code is complex. It is probably *too* complex for what it does.
I'm not sure what you would have me do about that, though.
I'll try to discuss this with you over chat.

Reply:
@fax4ever I pushed a few commits that replace the `IndexedEntityBindingContextProvider` with a `MappedIndexManagerFactory`. Overall it's more consistent, and a bit better than it was before this PR, which should be good enough considering improving the overall design of the bootstrap is out of scope.

In short, this:

```java
		PojoTypeAdditionalMetadata metadata = typeAdditionalMetadataProvider.get( indexedEntityType );
		// This metadata is guaranteed to exist; see prepareEntityOrIndexedType()
		PojoIndexedTypeAdditionalMetadata indexedTypeMetadata = metadata.getIndexedTypeMetadata().get();
		PojoEntityTypeAdditionalMetadata entityTypeMetadata = metadata.getEntityTypeMetadata().get();

		IndexManagerBuildingState<?> indexManagerBuildingState =
				contextProvider.getIndexManagerBuildingState(
						indexedTypeMetadata.getBackendName(),
						indexedTypeMetadata.getIndexName()
								.orElse( entityTypeMetadata.getEntityName() ),
						multiTenancyEnabled
				);
		IndexedEntityBindingContext bindingContext = contextProvider.createIndexedEntityBindingContext(
				this, indexManagerBuildingState.getSchemaRootNodeBuilder()
		);
		PojoIndexedTypeManagerBuilder<?, ?> builder = createIndexedTypeManagerBuilder(
				indexedEntityType, indexManagerBuildingState, bindingContext
		);
		// Put the builder in the map before anything else, so it will be closed on error
		indexedTypeManagerBuilders.put( indexedEntityType, builder );

		PojoMappingCollectorTypeNode collector = builder.asCollector();

		contributorProvider.get( indexedEntityType )
				.forEach( c -> c.contributeMapping( collector ) );
```

... became this:

```java
	private void mapIndexedType(PojoRawTypeModel<?> indexedEntityType,
			MappedIndexManagerFactory indexManagerFactory) {
		PojoTypeAdditionalMetadata metadata = typeAdditionalMetadataProvider.get( indexedEntityType );
		// This metadata is guaranteed to exist; see prepareEntityOrIndexedType()
		PojoIndexedTypeAdditionalMetadata indexedTypeMetadata = metadata.getIndexedTypeMetadata().get();
		PojoEntityTypeAdditionalMetadata entityTypeMetadata = metadata.getEntityTypeMetadata().get();

		MappedIndexManagerBuilder<?> indexManagerBuilder = indexManagerFactory.createMappedIndexManager(
				this,
				indexedTypeMetadata.getBackendName(),
				indexedTypeMetadata.getIndexName()
						.orElse( entityTypeMetadata.getEntityName() ),
				multiTenancyEnabled
		);
		PojoIndexedTypeManagerBuilder<?, ?> builder = createIndexedTypeManagerBuilder(
				indexedEntityType, indexManagerBuilder
		);
		// Put the builder in the map before anything else, so it will be closed on error
		indexedTypeManagerBuilders.put( indexedEntityType, builder );

		PojoMappingCollectorTypeNode collector = builder.asCollector();

		contributorProvider.get( indexedEntityType )
				.forEach( c -> c.contributeMapping( collector ) );
	}
```

So `IndexManagerBuildingState` and `IndexedEntityBindingContext` were merged into a single component, `MappedIndexManagerBuilder`. The `PojoIndexedTypeManagerBuilder` will have to collaborate with this component to create a type manager.

Does this look good (enough) to you?

Reply:
Thank you @yrodiere. Let me check...

Reply:
IMHO, it looks much better now.